solr/core/src/java/org/apache/solr/handler/component/QueryComponent.java

```diff
@@ -832,20 +832,24 @@ protected void createMainQuery(ResponseBuilder rb) {
     // perhaps we shouldn't attempt to parse the query at this level?
     // Alternate Idea: instead of specifying all these things at the upper level,
     // we could just specify that this is a shard request.
+    int shardRows;
     if (rb.shards_rows > -1) {
       // if the client set shards.rows set this explicity
-      sreq.params.set(CommonParams.ROWS, rb.shards_rows);
+      shardRows = rb.shards_rows;
     } else {
-      // what if rows<0 as it is allowed for grouped request??
-      sreq.params.set(
-          CommonParams.ROWS, rb.getSortSpec().getOffset() + rb.getSortSpec().getCount());
+      shardRows = rb.getSortSpec().getCount();
+      if (shardRows > 0) {
+        // If rows = -1 (grouped requests) or rows = 0, then there is no need to add the offset.
```

Comment:
the comment is nice but its placement here within this if block is confusing.  It applies to the *opposite*.  Please place it before.